File: src/lib/skills.ts
import { invoke } from "@tauri-apps/api/core";

export type SkillSource = "bundled" | "user";

export type SkillListItem = {
  name: string;
  description: string;
  source: SkillSource;
  enabled: boolean;
  error: string | null;
};

export type SkillMeta = {
  name: string;
  description: string;
  source: SkillSource;
};

export function skillsList(): Promise<SkillListItem[]> {
  return invoke<SkillListItem[]>("skills_list");
}

export function skillsSetEnabled(source: SkillSource, name: string, enabled: boolean): Promise<void> {
  return invoke("skills_set_enabled", { source, name, enabled });
}

/** Read-only preview: the files of one skill, `SKILL.md` first. Works for
 * disabled and broken skills too, unlike the assistant's `skill` tool. */
export function skillsFiles(source: SkillSource, name: string): Promise<string[]> {
  return invoke<string[]>("skills_files", { source, name });
}

export function skillsReadFile(source: SkillSource, name: string, path: string): Promise<string> {
  return invoke<string>("skills_read_file", { source, name, path });
}

export function skillsImport(path: string): Promise<SkillMeta> {
  return invoke<SkillMeta>("skills_import", { path });
}

export function skillsRemove(name: string): Promise<void> {
  return invoke("skills_remove", { name });
}

export function skillsUserDir(): Promise<string> {
  return invoke<string>("skills_user_dir");
}

/** Splits a `SKILL.md` into its YAML frontmatter and the Markdown after it —
 * the same fences `domain::agent_skills::split_frontmatter` parses. Without
 * this the viewer would render `name:`/`description:` as a setext heading.
 * `frontmatter` is `null` when there is none; `body` is then the whole file. */
export function splitSkillFrontmatter(content: string): {
  frontmatter: string | null;
  body: string;
} {
  const text = content.replace(/^\uFEFF/, "");
  const match = /^---[ \t]*\r?\n([\s\S]*?)\r?\n---[ \t]*(?:\r?\n|$)/.exec(text);
  if (!match) return { frontmatter: null, body: content };
  return { frontmatter: match[1], body: text.slice(match[0].length) };
}

File: src-tauri/src/commands/skills.rs
use crate::domain::agent_skills::{SkillListItem, SkillMeta, SkillSource};
use crate::infra::user_skills_store;
use crate::services::agent_skills;

#[tauri::command]
pub fn skills_list() -> Result<Vec<SkillListItem>, String> {
    agent_skills::list_skills().map_err(|e| e.to_string())
}

#[tauri::command]
pub fn skills_set_enabled(source: SkillSource, name: String, enabled: bool) -> Result<(), String> {
    agent_skills::set_skill_enabled(source, &name, enabled).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn skills_files(source: SkillSource, name: String) -> Result<Vec<String>, String> {
    agent_skills::preview_files(source, &name).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn skills_read_file(source: SkillSource, name: String, path: String) -> Result<String, String> {
    agent_skills::preview_file(source, &name, &path).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn skills_import(path: String) -> Result<SkillMeta, String> {
    user_skills_store::import_skill_dir(std::path::Path::new(&path)).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn skills_remove(name: String) -> Result<(), String> {
    agent_skills::remove_skill(&name).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn skills_user_dir() -> Result<String, String> {
    let dir = user_skills_store::ensure_user_skills_dir().map_err(|e| e.to_string())?;
    Ok(dir.to_string_lossy().into_owned())
}
